truffleruby/src/main/java/org/truffleruby/core/binding/BindingNodes.java

```diff
@@ -247,13 +236,15 @@ public Object localVariableGetCached(DynamicObject binding, String name,
 
         @Specialization(guards = {
                 "name == cachedName",
-                "!hiddenVariable(name)",
+                "!hiddenVariable(cachedName)",
                 "cachedFrameSlot != null",
+                "cachedMainFrameSlot == null",
                 "getExtrasDescriptorOrNull(binding) == descriptor"
         }, limit = "getCacheLimit()")
         public Object extraVariableGetCached(DynamicObject binding, String name,
                 @Cached("name") String cachedName,
                 @Cached("getExtrasDescriptorOrNull(binding)") FrameDescriptor descriptor,
+                @Cached("findFrameSlotOrNull(binding, name)") FrameSlotAndDepth cachedMainFrameSlot,
                 @Cached("findExtrasSlotOrNull(binding, name)") FrameSlotAndDepth cachedFrameSlot,
                 @Cached("createReadNode(cachedFrameSlot)") ReadFrameSlotNode readLocalVariableNode) {
             final MaterializedFrame frame = RubyArguments.getDeclarationFrame(Layouts.BINDING.getExtras(binding),
```

Comment:
If you have cachedFrameSlot != null and cachedMainFrameSlot == null you have the guarantee it's only in the extra frame, so no need to use `getDeclarationFrame` frame here.